contributors/design-proposals/custom-metrics-api.md

```diff
@@ -0,0 +1,302 @@
+metrics API.  Implementers *should* attempt to return all metrics with roughly
+identical timestamps and windows (when appropriate), but consumers should also
+verify that any differences in timestamps are within tolerances for
+a particular application (e.g. a dashboard might simply display the older
+metric with a note, while the horizontal pod autoscaler controller might choose
+to pretend it did not receive that metric value).
+
+### Labeled Metrics (or lack thereof) ###
+
+For metrics systems that support differentiating metrics beyond the Kubernetes
+object hierarchy (such as using additional labels), the metrics systems should
+have a metric which represents all such series aggregated together.
+Additionally, implementors may choose to the individual "sub-metrics" via the
+metric name, but this is expected to be fairly rare, since it most likely
+requires specific knowledge of individual metrics.  For instance, suppose we
+record filesystem usage by filesystem inside the container. There should then
+be a metric `filesystem/usage`, and the implementors of the API may choose to
+expose more detailed metrics like `filesystem/usage/my-first-filesystem`.
+
+Relationship to HPA v2
+----------------------
+
+The URL paths in this API are designed to correspond to different source
+types in the [HPA v2](hpa-v2.md).  Specifially, the `pods` source type
+corresponds to a URL of the form
+`/namespaces/$NS/pods/*/$METRIC_NAME?labelSelector=foo`, while the
+`object` source type corresponds to a URL of the form
+`/namespaces/$NS/$KIND.$GROUP/$OBJECT_NAME/$METRIC_NAME`.
+
+The HPA then takes the results, aggregates them together (in the case of
+the former source type), and uses the resulting value to produce a usage
+ratio.
+
+The resource source type is taken from the API provided by the
+"metrics" API group (the master/resource metrics API).
+
+The HPA will consume the API as a federated API server.
+
+Relationship to Resource Metrics API
+------------------------------------
+
+The metrics presented by this API may be a superset of those present in the
+resource metrics API, but this is not guaranteed.  Clients that need the
+information in the resource metrics API should use that to retrieve those
+metrics, and supplement those metrics with this API.
+
+Mechanical Concerns
+-------------------
+
+This API is intended to be implemented by monitoring pipelines (e.g.
+inside Heapster, or as an adapter on top of a solution like Prometheus).
+It shares many mechanical requirements with normal Kubernetes APIs, such
+as needed to support encoding different versions of objects in both JSON
+and protobuf, as well as acting as a discoverable API server.  For these
+reasons, it is expected that implemenators will make use of the Kubernetes
+genericapiserver code.  If implementors choose not to use this, they must
+still follow all of the Kubernetes API server conventions in order to work
+properly with consumers of the API.
```

Comment:
make this more specific, especially around which API verbs are supported (appears only to be get?) and what semantics resourceVersion carries for MetricValue objects